### ISSUE TITLE
JSON errors found by the containers test script

### ISSUE THREAD

**Issue by akirashirosawa** [CleverRaven/Cataclysm-DDA#43149]
# Describe the bug 
After I added a `debugmsg` to the `item::put_in` function (PR #41677), I found a lot of bugs related to containers. The part related to the C++ code I tried to fix right away, in the PR #41677. To fix the errors related to JSON https://github.com/CleverRaven/Cataclysm-DDA/pull/41677#issuecomment-670510332 , I promised to write a script and did it (the script turned out to be much more complicated than I expected). The script is written in nodejs and validates the **JSON** related with **items**, **containers**, **item_groups**, **professions**. I implemented the logic from the docs, however, some things are not documented, and I had to look at the C++ code (i will try to point out these points).
Script gist [here](https://gist.github.com/akirashirosawa/1def3bede211fa3183babb450c5f93b8). Script output:
#### 1. Raw JSON read
- [x] 1.1 JSON in file not array
```
WARNING: JSON in the file (data/json/itemgroups/altered_item.json) is not array
WARNING: JSON in the file (data/json/snippets/nightmares.json) is not array
```
The PR #41498 make all JSON in files was in arrays for ease of handling (in`jq` etc). As you can see, there is degradation in these two files.

- [x] 1.2 Duplicate identifier
```
ERROR: JSON object (laptop) type (item) with duplicate id or abstract
```
On the version of Cataclysm for which I wrote the script, there were much more problems with the duplicate identifier, but many were fixed in this PR `Deduplicating JSON objects` #42989. However, only object identifiers of strictly the same type have been corrected. In my script the `item` type is "composite", so it caught this problem. First laptop "GENERIC", "A broken laptop, basically a paperweight now." Second: "TOOL", "A laptop computer with replaceable batteries." Only the second is visible in the game (in item spawn menu). Delete the first one completely or rename?

#### 2. Items build (resolve copy-from)
- [x] 2.1 `price_post` issues
```
WARNING: relative undefined: item (book_nonf_hard_phil_tpl) relative price_post (-2500)
WARNING: relative undefined: item (book_nonf_soft_phil_tpl) relative price_post (-2600)
WARNING: relative undefined: item (paperback_romance_spb) relative price_post (-800)
WARNING: relative undefined: item (paperback_romance_lnl) relative price_post (-400)
WARNING: relative undefined: item (paperback_romance_vanilla) relative price_post (-100)
WARNING: relative undefined: item (paperback_western_em) relative price_post (-100)
WARNING: relative undefined: item (book_fict_soft_ya_vanilla) relative price_post (-400)
WARNING: relative undefined: item (book_fict_soft_ya_quiddity) relative price_post (-400)
WARNING: relative undefined: item (book_fict_hard_ya_quiddity) relative price_post (-400)
WARNING: relative undefined: item (book_archery) relative price_post (2500)
WARNING: relative undefined: item (book_nonf_hard_dodge_tlwd) relative price_post (2000)
```
Both JSON and C++ code used `"price_postapoc"`. In `Item_factory :: load_basic_info` only `"price_postapoc"` JSON key is handled: `assign (jo, "price_postapoc", def.price_post, false, 0_cent); `. So `"price_post"` is an error in JSON and in the DOCS(!) `doc/JSON_INFO.md`: `" price_post ":" 1 USD "`. It is need to change in the docs and in JSON `"price_post"` to `"price_postapoc"`, there will be `11` replacements + `1` in docs. Or leave it as in the docs and change `"price_postapoc"` to `"price_post"` there will be `3960` replacements + `1` in cpp.

- [x] 2.2 proportional price
```
WARNING: proportional undefined: item (mutant_bug_lungs) proportional price (0.1)
```
Everything is simple here: the parent (lung) of this item (mutant_bug_lungs) does not have the `price` key, only `price_postapoc`, so this does not work in the game (despite the fact that `price` = `price_postapoc` by default). I can attach screenshots in confirmation. Need to change `mutant_bug_lungs` `proportional` `price` to `price_postapoc`.
#### 3. Itemgroups build
- [x] Everything is fine here, except that some entries contain the key `sealed: true | false`, which is not documented. All itemgroup entries match the schema (Can see `itemgroupEntrySchema` in the script).
#### 4. Item test
- [x] Checks if the `item` can spawn in its default `container`.
```
ERROR: item (nitrox) container (bottle_plastic) pocket not found, because:
       pocket(0) no airtight for gas
ERROR: item (extinguishing_agent) container (bottle_plastic) pocket not found, because:
       pocket(0) no airtight for gas
ERROR: item (fertilizer_commercial) container (bag_canvas) pocket not found, because:
       max_contains_weight (15000) < weight (21000) count (60)
ERROR: item (offal_canned) container (can_food) pocket not found, because:
       max_contains_volume (250) < volume (500)
ERROR: item (cereal2) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (toastem) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (toastem2) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (toastem3) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (candycigarette) container (box_cigarette) pocket not found, because:
       max_contains_volume (85) < volume (90)
ERROR: item (frozen_dinner) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (fchicken) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (pie_meat) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (2000)
ERROR: item (pizza_meat) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (pizza_supreme) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (pelmeni) container (can_food) pocket not found, because:
       max_contains_volume (250) < volume (500)
ERROR: item (mutagen_jabberblood) container (flask_glass) pocket not found, because:
       max_contains_weight (1000) < weight (1250) count (5)
ERROR: item (hickory_nut_ambrosia) container (bottle_plastic) pocket not found, because:
       max_contains_weight (1000) < weight (1016) count (4)
ERROR: item (pie_veggy) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (2000)
ERROR: item (pizza_veggy) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (crackers) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (pie) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (2000)
ERROR: item (pizza_cheese) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
ERROR: item (pie_maple) container (box_small) pocket not found, because:
       max_contains_volume (990) < volume (1000)
```
Need to change `box_small` `max_contains_volume` from `990 ml` to `1L`. And change its `volume` form `1 L` to `1010 ml` if need `volume` > `max_contains_volume`,  `bottle_plastic` pocket add `airtight`, etc.
#### 5. Itemgroups test
Checks if the `item_group` can spawn `item` with specified `charges`, `container-group`, `contents-group`, `ammo-group`, `contents-item`, `container-item`, `ammo-item`.

- [x] 5.1 `can't contain charges` aka `tried to set ammo for spray_can which does not have ammo or a magazine`
```
ERROR: item spray_can itemgroup (recycle_aluminum) can't contain charges (3)
ERROR: item jumper_cable_heavy itemgroup (recycle_steel) can't contain charges (10)
ERROR: item jumper_cable itemgroup (recycle_steel) can't contain charges (2)
ERROR: item permanent_marker itemgroup (SUS_junk_drawer) can't contain charges (50)
ERROR: item permanent_marker itemgroup (SUS_junk_drawer) can't contain charges (50)
ERROR: item candle itemgroup (SUS_junk_drawer) can't contain charges (10)
ERROR: item permanent_marker itemgroup (SUS_junk_drawer_artsy) can't contain charges (70)
ERROR: item permanent_marker itemgroup (SUS_junk_drawer_handy) can't contain charges (50)
ERROR: item candle itemgroup (SUS_junk_drawer_tidy) can't contain charges (70)
ERROR: item tanbark itemgroup (teashop_bulk_teas) can't contain charges (2)
ERROR: item spurge itemgroup (teashop_bulk_teas) can't contain charges (2)
ERROR: item lotus itemgroup (teashop_bulk_teas) can't contain charges (2)
ERROR: item chamomile itemgroup (teashop_bulk_teas) can't contain charges (2)
```
The problem is in this code:
```cpp
// item_group.cpp
// void Item_modifier::modify( item &new_item ) const
    if( ch != -1 ) {
        if( new_item.count_by_charges() || new_item.made_of( phase_id::LIQUID ) ) {
            new_item.charges = std::max( 1, ch );
        } else if( new_item.is_tool() ) {
            if( !new_item.magazine_default().is_null() ) {
                item mag( new_item.magazine_default() );
                mag.ammo_set( mag.ammo_default(), ch );
                new_item.put_in( mag, item_pocket::pocket_type::MAGAZINE_WELL );
            } else if( new_item.is_magazine() ) {
                new_item.ammo_set( new_item.ammo_default(), ch );
            } else {
                debugmsg( "tried to set ammo for %s which does not have ammo or a magazine",
                          new_item.typeId().c_str() );
            }
        } else if( new_item.type->can_have_charges() ) {
            new_item.charges = ch;
        }
    }
```
As you can see that this code can only place `charges` in `items` with `pockets` (`MAGAZINE_WELL` or `MAGAZINE`). That is, in fact, at the moment, an `item` without `pockets` cannot be charged through the `itemgroup`. Need to fix JSON `itemgroups` by removing `charges`, or add `pockets` to JSON `items`, or rewrite C++ code to make it can handle items without `pockets` (well, which are not `count_by_charges` or `liquid`).

- [x] 5.2 ammo_restriction < charges
This is not an error, since C++ code will reduce the number of charges to an acceptable. However, there is probably no point in trying to shove 20 rounds into an M24 when it only holds 5: `{ "item": "M24", "prob": 15, "charges-min": 0, "charges-max": 20 }`
```
WARNING: item (p226mag_12rd_357sig) itemgroup (guns_pistol_common) ammo_restriction (357sig_jhp) count (12) < charges (15)
WARNING: item (p320mag_13rd_357sig) itemgroup (guns_pistol_common) ammo_restriction (357sig_jhp) count (13) < charges (14)
WARNING: item (bhp9mag_13rd) itemgroup (guns_pistol_common) ammo_restriction (9mm) count (13) < charges (15)
WARNING: item (ppq9mag_10rd) itemgroup (guns_pistol_common) ammo_restriction (9mm) count (10) < charges (17)
WARNING: item (ppq40mag_10rd) itemgroup (guns_pistol_common) ammo_restriction (40sw) count (10) < charges (14)
WARNING: item (hptc9mag_8rd) itemgroup (guns_pistol_common) ammo_restriction (9mm) count (8) < charges (15)
WARNING: item (hptcf380mag_8rd) itemgroup (guns_pistol_common) ammo_restriction (380_JHP) count (8) < charges (10)
WARNING: item (cz75mag_12rd) itemgroup (guns_pistol_common) ammo_restriction (9mm) count (12) < charges (26)
WARNING: item (glock_20mag) itemgroup (guns_pistol_rare) ammo_restriction (10mm_fmj) count (15) < charges (30)
WARNING: item (1895sbl) itemgroup (guns_rifle_common) ammo_restriction (4570_sp) count (6) < charges (7)
WARNING: item (M24) itemgroup (guns_rifle_rare) ammo_restriction (762_51) count (5) < charges (20)
WARNING: item (famasmag) itemgroup (guns_rifle_obscure) ammo_restriction (556) count (25) < charges (30)
WARNING: item (remington_870) itemgroup (guns_shotgun_common) ammo_restriction (shot_00) count (5) < charges (6)
WARNING: item (remington_1100) itemgroup (guns_shotgun_common) ammo_restriction (shot_00) count (5) < charges (10)
WARNING: item (tavor_12) itemgroup (guns_shotgun_rare) ammo_restriction (shot_00) count (5) < charges (16)
WARNING: item (mossberg_500) itemgroup (longguns_cop) ammo_restriction (shot_00) count (6) < charges (8)
WARNING: item (remington_870) itemgroup (longguns_cop) ammo_restriction (shot_00) count (5) < charges (6)
WARNING: item (mossberg_500) itemgroup (guns_survival) ammo_restriction (shot_00) count (6) < charges (8)
WARNING: item (remington_870) itemgroup (guns_survival) ammo_restriction (shot_00) count (5) < charges (6)
WARNING: item (storage_battery) itemgroup (gas_charging_rack) ammo_restriction (battery) count (40000) < charges (100000)
WARNING: item (small_storage_battery) itemgroup (gas_charging_rack) ammo_restriction (battery) count (500) < charges (1000)
WARNING: item (medium_storage_battery) itemgroup (gas_charging_rack) ammo_restriction (battery) count (7000) < charges (10000)
WARNING: item (mossberg_500) itemgroup (survivor_longguns) ammo_restriction (shot_00) count (6) < charges (8)
WARNING: item (remington_870) itemgroup (survivor_longguns) ammo_restriction (shot_00) count (5) < charges (6)
WARNING: item (glock17_17) itemgroup (NC_robofac_generic_security_misc) ammo_restriction (9mm) count (17) < charges (21)
WARNING: item (glock17_17) itemgroup (NC_robofac_generic_security_misc) ammo_restriction (9mmP) count (17) < charges (21)
WARNING: item (M24) itemgroup (prof_sniper_police_both) ammo_restriction (762_51) count (5) < charges (20)

```
Can be fixed, can be ignored.

- [ ] 5.3 Container not fit
```
ERROR: item (celery) itemgroup (salad_bar) container (bowl_plastic) pocket not found, because:
       max_contains_volume (750) < volume (1000)
ERROR: item (cabbage) itemgroup (salad_bar) container (bowl_plastic) pocket not found, because:
       max_contains_volume (750) < volume (2250)
       max_contains_weight (1000) < weight (1251) count (9)
       max_item_length (128) < longest_side (131)
ERROR: item (m4a1) itemgroup (prof_rescuer_both) container (back_holster) pocket not found, because:
       max_item_length (600) < longest_side (758)
ERROR: item (M24) itemgroup (prof_sniper_police_both) container (back_holster) pocket not found, because:
       max_item_length (600) < longest_side (992)
```

- [ ] 5.4  `ammo-item` not `ammo`
```
WARNING: itemgroup (prof_camper_both) ammo-item (water_clean) type not AMMO or MAGAZINE
ERROR: item (water_clean) itemgroup (prof_camper_both) container (water_clean) not have pocket_data
```
Probably a typo. `water_clean` try to put into `water_clean`. `prof_camper_both` means profession camper both sexes.
```json
"type": "profession",
"id": "camper",
...
{ "item": "water_clean", "ammo-item": "water_clean", "container-item": "canteen" },
```
Fix: `{ "item": "water_clean", "container-item": "canteen" },` do the same.
# Steps To Reproduce
download gist and `./tools/test_contaiters.js`

# Expected behavior
No `ERROR`s, if possible without `WARNING`s.

# Versions and configuration
DDA Version last commit 2eef30cd1cfd4fa1fab9fb4b4ceeb4284281ae1e

**Comment by jbytheway** [comment]
After some discussion, I am leaning towards making jumper cables a container of some semi-abstract "cable spool" item (similar to our "battery" item).  I think that would require the least code changes, because that's how most charges-based items have been migrated, so that path should be relatively smooth.

Testing that jumper cables are working correctly is messy, though.  They can interact with vehicles, movement, and bionics, so there are a lot of corner cases to worry about (e.g. getting smashed across the room by a hulk while unspooling a cable).

**Comment by misterprimus** [comment]
Seems to have fixed guns, but gas masks and other items with charges like lighters still only spawn on dead zeds with either 100% or 0% charges, never anything else.

**Comment by akirashirosawa** [comment]
@misterprimus 
> Seems to have fixed guns, but gas masks and other items with charges like lighters still only spawn on dead zeds with either 100% or 0% charges, never anything else.

This issue is only about errors in JSON. There are many more bugs in C++ regarding containers. They are being fixed here PR #41677 but have not been merged yet. Try to write comment in #41677, and ask for PR to be merged faster :) It contains many containers/pockets fixes.
